### PR TITLE
Add nvFuser fusion region repro util and update DebugTransform

### PR DESCRIPTION
**Pull request by riccardofelluga**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

From discussions in #387 and #783, this PR adds two helper functions:
- One to print a repro script for a specific nvFuser segment; and
- Another to get a fusion definition from the trace(avoids manually querying the context).

I also added a small section in the documentation to explain the usage of the helper function.

Also in this PR an update to the DebugTransform such that it is possible to debug symbols before and after any interesting section of the trace. I don't really like that transform since it's an hack but for now it does the job. I think in the future we should consider adding a debug prim to intercept inputs and add information to a trace.

## Usage:
As described in the documentation to use this feature you need two things:

1. Add the `nv_store_fusion_inputs=True` compile option; and
2. Call `get_nvfuser_repro` with the trace and the name of the fusion you want to see the repro of.

Note: ~This PR depends on nvFuser PR [#2732](https://github.com/NVIDIA/Fuser/pull/2732). Hence I will convert this PR to not a draft when that one passes.~ MERGED!

### CODE REVIEW

File: docs/source/basic/inspecting_traces.rst [hunk 1]
```diff
@@ -189,6 +189,46 @@ which prints::
     }
   }
 
+Moreover, if you are just interested in running a specific nvFuser region without Thunder, you can use an handy helper function. The ``get_nvfuser_repro()`` function takes a trace and a fusion name as input and returns it's repro script::
```
Comment (crcrpar) [inline docs/source/basic/inspecting_traces.rst:192]:
```suggestion
Moreover, if you are just interested in running a specific nvFuser region without Thunder, you can use a handy helper function. The ``get_nvfuser_repro()`` function takes a trace and a fusion name as input and returns it's repro script::
```

File: thunder/executors/nvfuserex_impl.py [hunk 2]
```diff
@@ -479,6 +484,9 @@ def create_fusion_definition_wrapper(
     #   objects is captured by names in the trace.
     # These properties are distinct from the inputs and outputs to the trace itself, which
     #   may contain duplicates and whose order must be preserved.
+    store_inputs: None | bool = get_compile_option(
+        "nv_store_fusion_inputs", "Enable nvFuser store fusion inputs for repro."
```
Comment (crcrpar) [inline thunder/executors/nvfuserex_impl.py:488]: "Enable nvFuser to store fusion inputs for repro" or "Let nvFuser store fusion inputs for repro"?

File: thunder/examine/__init__.py [hunk 3]
```diff
@@ -216,3 +216,41 @@ def get_fusion_symbols(trace: TraceCtx, warn_if_fusions_unavailable: bool = True
...
+
+    if fusion.last_used is None:
+        assert (
+            False
+        ), "Fusion definition needs to be executed to record the inputs. You must execute the fusion first before you can query the repro."
+
+    if fusion.last_inputs is None:
+        assert (
+            False
+        ), "Fusion definition inputs need to be recorded. Use compile option 'nv_store_fusion_inputs=True' while tracing."
```
Comment (crcrpar) [inline thunder/examine/__init__.py:253]: why using `assert`, rather than raise?

File: thunder/executors/nvfuserex_impl.py (hunk 2 above)
Comment (riccardofelluga) [reply thunder/executors/nvfuserex_impl.py:488]: Yeah I red the descriptions of the other nv compile options and I went for something similar but it better to phrase it differently

File: thunder/examine/__init__.py (hunk 3 above)
Comment (riccardofelluga) [reply thunder/examine/__init__.py:253]: There is not particular reason, let me switch to raise

**Review by kshitij12345 — APPROVED** [review]
LGTM @riccardofelluga, thank you!

Only question I had was - do you think we should have a small test for `get_nvfuser_repro`?

**Comment by riccardofelluga** [comment]
Thanks @kshitij12345 ! It is true that there is no thing such as excess of caution 😁 however since the method relies on a feature from nvFuser it might be risky to test it against an hardcoded result because it could change on their side.

**Review by t-vi — APPROVED** [review]
Thank you @riccardofelluga @kshitij12345 @crcrpar

**Pull request merged**